Refactoring and few tests to explain the logic

byapparov opened the pull request:
* Added assert_that instead of stopifnot

* Refactored the cbc_solve function - externalised parameter transformation.

byapparov committed:
* Added assert_that insead of stopifnot.

* Refactored the cbc_solve function - externalised parameter transformation.

byapparov committed:
Updated description file.

byapparov committed:
Fix build - example failure.

dirkschumacher commented:
Thanks!

dirkschumacher commented on DESCRIPTION:
```diff
@@ -15,6 +15,7 @@ LinkingTo: Rcpp
     methods, 
-    Matrix
+    Matrix,
+    assertthat
```
I would prefer not to have another dependency for argument checking. All those `assert_that` can also be modelled by `stopifnot`.  Would be great if you could revert that `asserthat` change.

What is better though, is to give a user better error messages instead of an expression that failed. E.g.
` if (!is.numeric(obj)) { stop("'obj' is not a numeric vector", call. = FALSE) }`
But that can also be a different PR.

byapparov replied:
assertthat package is used in every package Hadley W touches (which is a lot), so I would not hesitate much in using it as a dependancy.

it provides messages you want for free:

``` R
library(assertthat)
x <- "c"
assert_that(is.numeric(x))
# Error: x is not a numeric or integer vector
```

dirkschumacher replied:
I looked at the reverse dependencies and it seems that not that many of the tidyverse package use the package. But fine, let's pull in that dependency.

dirkschumacher commented on R/cbc_solve.R:
```diff
@@ -37,42 +39,64 @@ cbc_solve <- function(obj,
+#'
+#' @noRd
+#' @param ... parameters for the cbc optimisation
+prepare_cbc_args <- function(...) {
```
Why `...` and not just a single argument that is a list? Because in line 57 you convert the list to multiple arguments and then in line 80 back to a list. You could pass `cbc_args` directly.

byapparov replied:
I could, but it makes it easier to test and reason (for me). See tests, this function abstracts from the implementation of `cbc_solve`. The fact that we need to convert the list two times is a side effect of `cbc_solve` implementation.

We should not care if variables are converted back and forth as soon as abstraction of the individual units is correct.

dirkschumacher replied:
It just makes it a bit harder to program against that interface and simply passing a named list. But it is also just an internal function, to why not.

dirkschumacher commented on R/cbc_solve.R:
```diff
@@ -37,42 +39,64 @@ cbc_solve <- function(obj,
+prepare_cbc_args <- function(...) {
+  cbc_args <- list(...)
+
+  if (length(cbc_args) == 0) {
```
Instead of `0`, could you please write `0L`. Also for the other calls that have hardcoded integers

dirkschumacher commented on R/cbc_solve.R:
```diff
@@ -37,42 +39,64 @@ cbc_solve <- function(obj,
+  names(cbc_args)[has_name] <- paste0("-", names(cbc_args)[has_name])
+  cbc_args[!has_name] <- paste0("-", cbc_args[!has_name])
+
+  return(cbc_args)
```
Please remove the `return`

dirkschumacher reviewed (changes requested):
Please also add yourself as a contributor `role="ctb"` to the `DESCRIPTION` file - if you want. I assume that your code falls under the same license.

byapparov committed:
Remove return and integerise zeros

byapparov committed:
Added self as contributor to DESCRIPTION.

byapparov committed:
A bit more refactoring in response to return() comment.

byapparov committed:
Final purification of the args calculation.

byapparov committed:
Moving the construction of the parameter list to R to make it testable.

This reduces amount of cpp code in a single function. As test seem to have good coverage of different parameter calls, this should not be a problem once build passes.

byapparov committed:
Small documentation change to trigger travis build.

byapparov committed:
modeling args without values correctly.

byapparov committed:
Refactor tests to follow BDD style of previous tests.

byapparov committed:
typo fix

byapparov commented:
@dirkschumacher I know I got a bit of road with this PR.  

Hope these changes help as amount of cpp code is reduced and tests are added.

dirkschumacher commented:
Thanks. Can I start reviewing it or do you plan to add more changes to the PR?

dirkschumacher commented on R/cbc_solve.R:
```diff
@@ -1,3 +1,5 @@
+#' @import assertthat
```
Could you please use `@imprtFrom` and only import the functions actually use? `assert_that`, `is.numeric`

dirkschumacher commented on R/cbc_solve.R:
```diff
@@ -37,44 +39,94 @@ cbc_solve <- function(obj,
+# Appends prefix to argument names with value
+# @noRd
+rename_cbc_args <- function(names, values) {
+  if (length(values) == 0)
```
Can you add curly brackets and replace `0` by `0L`?

byapparov committed:
Resonse to code review:

* 0 changed to 0L

* {} added to all if() statements

* Use @importFrom (note: is.numeric is part of base)

byapparov commented:
@dirkschumacher thank you for review, I have made requested changes. 

I have finished working on this PR.
